File: solutions/cpp/Search_BST_for_first_occurrence_recursive.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <memory>

#include "./BST_prototype.h"

using std::unique_ptr;

// @include
BSTNode<int>* FindFirstEqualK(const unique_ptr<BSTNode<int>>& tree, int k) {
  if (!tree) {
    return nullptr;  // No match.
  } else if (tree->data == k) {
    // Recursively search the left subtree for first node containing k.
    auto* node = FindFirstEqualK(tree->left, k);
    return node ? node : tree.get();
  }
  // Search the left or right subtree based on relative values of tree->data
  // and k.
  return FindFirstEqualK(tree->data < k ? tree->right : tree->left, k);
}
// @exclude

int main(int argc, char* argv[]) {
  //    3
  //  2   6
  // 1   4 6
  auto root = unique_ptr<BSTNode<int>>(new BSTNode<int>{3});
  root->left = unique_ptr<BSTNode<int>>(new BSTNode<int>{2});
  root->left->left = unique_ptr<BSTNode<int>>(new BSTNode<int>{1});
  root->right = unique_ptr<BSTNode<int>>(new BSTNode<int>{6});
  root->right->left = unique_ptr<BSTNode<int>>(new BSTNode<int>{4});
  root->right->right = unique_ptr<BSTNode<int>>(new BSTNode<int>{6});
  assert(!FindFirstEqualK(root, 7));
  assert(FindFirstEqualK(root, 6)->data == 6 &&
         FindFirstEqualK(root, 6)->right->data == 6);

  //    3
  //  3   5
  // 2   5 6
  root = unique_ptr<BSTNode<int>>(new BSTNode<int>{3});
  root->left = unique_ptr<BSTNode<int>>(new BSTNode<int>{3});
  root->left->left = unique_ptr<BSTNode<int>>(new BSTNode<int>{2});
  root->right = unique_ptr<BSTNode<int>>(new BSTNode<int>{5});
  root->right->left = unique_ptr<BSTNode<int>>(new BSTNode<int>{5});
  root->right->right = unique_ptr<BSTNode<int>>(new BSTNode<int>{6});
  assert(!FindFirstEqualK(root, 7));
  assert(FindFirstEqualK(root, 3) == root->left.get());
  assert(FindFirstEqualK(root, 5) == root->right->left.get());
  assert(FindFirstEqualK(root, 6)->data == 6);
  return 0;
}

File: solutions/python/search_bst_for_first_occurrence_recursive.py
from binary_tree_prototype import BinaryTreeNode


# @include
def find_first_equal_k(tree, k):
    if not tree:
        return None  # No match.
    elif tree.data == k:
        # Recursively search the left subtree for first node containing k.
        node = find_first_equal_k(tree.left, k)
        return node or tree
    # Search the left or right subtree based on relative values of tree.data
    # and k.
    return find_first_equal_k(tree.right if tree.data < k else tree.left, k)
# @exclude


def main():
    #      3
    #    2   6
    #  1    4 6
    root = BinaryTreeNode(3)
    root.left = BinaryTreeNode(2)
    root.left.left = BinaryTreeNode(1)
    root.right = BinaryTreeNode(6)
    root.right.left = BinaryTreeNode(4)
    root.right.right = BinaryTreeNode(6)
    assert not find_first_equal_k(root, 7)
    assert (find_first_equal_k(root, 6).data == 6 and
            find_first_equal_k(root, 6).right.data == 6)

    #      3
    #    3   5
    #  2    5 6
    root = BinaryTreeNode(3)
    root.left = BinaryTreeNode(3)
    root.left.left = BinaryTreeNode(2)
    root.right = BinaryTreeNode(5)
    root.right.left = BinaryTreeNode(5)
    root.right.right = BinaryTreeNode(6)
    assert not find_first_equal_k(root, 7)
    assert find_first_equal_k(root, 3) is root.left
    assert find_first_equal_k(root, 5) is root.right.left
    assert find_first_equal_k(root, 6).data == 6


if __name__ == '__main__':
    main()
